g3doc/Tutorials/Toy/Ch-4.md

```diff
@@ -162,81 +152,62 @@ func @transpose_transpose(%arg0: !toy<"array">)
 
 Perfect! No `transpose` operation is left, the code is optimal.
 
-The code in `mlir/ToyCombine.cpp` implements a few more patterns that eliminate
-trivial reshapes, or fold them into constants.
 
-# Shape Inference and Generic Function Specialization
+# Optimize Reshapes
 
-Our IR operates on generic arrays, we don't know the shape of the arrays other
-than during initialization of constants. However we can propagate the shapes
-through the computation until they are all known. The issue is how to handle
-calls to user-defined generic functions: every call site could deduce different
-shapes. One possibility would be to perform symbolic inference based on the
-argument types, but this would be hard to generalize if we were to introduce
-more control flow in the language. Instead we will proceed by function
-specialization: for every call site with new argument shapes we duplicate the
-function and specialize it. This is akin to C++ template instantiation:
+TableGen also provides a method for adding argument constraints when the transformation 
```

Comment:
TableGen -> ODS

I believe this is the formal definition, we have been avoiding using TableGen in the messaging.

Reply:
Updated. Is there a reason for this?

Reply:
I think we are leaving the future open to using something other than tablegen to define the patterns, so hardcoding tablegen would make it more difficult to backtrack.

Reply:
Right, I do recall that discussion.